lib/mongoid/fields.rb

```diff
@@ -270,38 +270,98 @@ def option(option_name, &block)
+          # key. We can convert them back to their names by looking in the
+          # aliased_associations hash.
+          aliased = meth
+          if as && a = as.fetch(meth, nil)
+            aliased = a.to_s
+          end
 
-      # Returns an array of names for the attributes available on this object.
-      #
-      # Provides the field names in an ORM-agnostic way. Rails v3.1+ uses this
-      # method to automatically wrap params in JSON requests.
-      #
-      # @example Get the field names
-      #   Model.attribute_names
-      #
-      # @return [ Array<String> ] The field names
-      def attribute_names
-        fields.keys
+          field = nil
+          klass = nil
+          if fs && f = fs[aliased]
+            field = f
+            yield(meth, f, true) if block_given?
+          elsif rs && rel = rs[aliased]
+            klass = rel.klass
+            yield(meth, rel, false) if block_given?
+          else
+            yield(meth, nil, false) if block_given?
+          end
+        end
+        field
       end
 
       # Get the name of the provided field as it is stored in the database.
       # Used in determining if the field is aliased or not. Recursively
       # finds aliases for embedded documents and fields, delimited with
       # period "." character.
       #
-      # @example Get the database field name of a field.
-      #   Model.database_field_name(:authorization)
+      # This method will not expand the alias of a referenced association that
+      # is not the last item. For example, if we had a School that has_many
+      # Students, and the field name passed was:
+      #
+      #   school._id
+      #
+      # The alias for a referenced association is that association's _id field.
+      # Therefore, expanding out this association would yield:
+      #
+      #   school_id._id
       #
-      # @example Get the database field name of an embedded field.
-      #   Model.database_field_name('customers.addresses.city')
+      # This is not the correct field name, because the intention here was not
+      # to get a property of the _id field. The intention was to get a property
+      # of the referenced document. Therefore, if a part of the name passed is
+      # a referenced association that is not the last part of the name, we
+      # won't expand its alias, and return:
+      #
+      #   school._id
+      #
+      # If the referenced association is the last part of the name, we will
+      # pass back the _id field.
       #
       # @param [ String, Symbol ] name The name to get.
+      # @param [ Hash ] relations The associations.
+      # @param [ Hash ] alaiased_fields The aliased fields.
+      # @param [ Hash ] alaiased_associations The aliased associations.
       #
       # @return [ String ] The name of the field as stored in the database.
-      def database_field_name(name)
+      def database_field_name(name, relations, aliased_fields, aliased_associations)
```

Comment:
This is an api change, can this method be `@api private` as well?

Reply:
This is actually not an API change since I created a separate `database_field_name` with the old API that calls this

Reply:
done